### DISCUSSION TITLE
Aggregator Service Specification

### DISCUSSION THREAD

**Discussion by gimer** [cartesi/cips#31] [General]
Hi everyone,

I'm creating this document in order to start formalizing and discussing some concepts related to the aggregator service. This is work in progress.

**Aggregator Service Specification (Work in progress)**

The goal of the aggregator is to collect a batch of transactions and group them into a single message in order to improve scalability without compromising privacy and security. Aggregation is at the beginning centralized, however, an incentive market should exist in the future to encourage other nodes to become an aggregator.

What is signature aggregation?
A signature aggregation scheme is a mechanism to generate a single signature that is sufficient to authenticate multiple message and public keys, e.g., BLS signatures [1].

The aggregator node (AN) is the entry point and it has the following responsibilities: 

- Encode/decode the received data. 
- Verify the transactions’ signatures. 
- Aggregate the transactions’ data. 
- Aggregate the transactions’ signatures.
- Generate a secret and a public key to sign the transactions batch.
- Sign the aggregated transactions. 
- Push the signed message further into the network.

**Definitions**:

Signature Verification Process (SigVer): during this process the aggregator verifies the individual signatures of every transaction. 

Maximum Number of transactions (MNT): this value is the maximum number of transactions to be aggregated.

Collection Time Interval (CTI): It is a *fixed* time interval to collect transactions that will be aggregated.

Aggregation Process: during this period, the aggregator batch transactions into only one single message. It starts when either the maximum number of transactions has been received or the Collection Time Interval has ended.

Aggregate Algorithm: this algorithm aggregates multiple signatures into one single message.

**Questions**

How to collect and calculate fees? Do all transactions pay the same fee? 

Is the transaction’s arriving time the only parameter to order transactions (i.e., FIFO)?

If an aggregator is not a sequencer, could instant finality be guaranteed? I guess not.

Should we consider a minimum number of transactions to be aggregated?

Is it possible for the sender to cancel a transaction during the collection period?

Is there any restriction for a dApp/user to send transactions? Do we need an authentication process?

What happens if an aggregator goes offline?

What happens if the aggregator’s output is rejected?

References:

[1] [BLS Signatures Draft V4](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04).

**Comment by guidanoli** [comment]
> Should we consider a minimum number of transactions to be aggregated?

I think it is safe to say that we should aggregate at least 2 transactions, right?
Because to aggregate 0 transactions doesn't make sense, and to aggregate 1 transaction sounds counter-productive.

**Reply by guidanoli** [reply]
But this is just a lower bound, maybe the minimum number of transactions should be the smallest number of transactions that makes it cheaper to commit all transactions using the Aggregation Algorithm rather than committing each transaction individually.

**Reply by felipeargento** [reply]
I would agree with you from the financial point of view, in the sense that you'd be losing money by aggregating a number of transactions smaller than the break even. 

However, the user will either send his transaction to the aggregator (and get instant feedback) or to the  input box directly (in the blockchain). If an input is sent to the aggregator I'd say that it has to be included (respecting the aggregation block/instant feedback) regardless of the profitability of the aggregator. I'd go as far as suggesting that we might need a penalty for aggregators that promise they'll include a transaction and don't. 

So I don't think we should have a minimum. What we can have in the future is a strategy in which the aggregator uses different aggregation techniques depending on the amount of transactions he is going to include - maybe the benefits of each schema vary a lot with the amount of txs. But that is far in the future probably.

**Reply by guidanoli** [reply]
So, let me know if I got this right: if users want instant feedback from aggregators, they also need to take the risk of paying a little extra in the case where the aggregator does not receive enough transactions before the CTI to make a profit out of aggregation.

**Comment by guidanoli** [comment]
> How to collect and calculate fees? Do all transactions pay the same fee?

## Cost per transaction

I think it makes sense to calculate transaction fees by some base fee ($C_1$), by the computational complexity of processing them (which could be proportional to the size of the message, $C_2$) and by the cost of committing the aggregated transaction to the blockchain (which could be inversely proportional to the number of aggregated transactions, $C_3$). Let $n$ be the size of the message, and $m$, the number of messages.

$$ C_{tx}(n,m) = C_1 + n \cdot C_2 + \frac{1}{m} C_3 $$

Now, you see, the user won't know much it should pay when submitting its tx, because $m$ might change before the A.N. submits the aggregated transaction containing it. To mitigate this, we could develop some frontend to give users an estimation of how much they would pay based on current network traffic, like [ETH Gas Station](https://www.ethgasstation.info/).

## Cost per block

We can then calculate the total amount that an A.N. would receive from a block with $m>0$ messages. Let $n_i$ denote the size of the $i$-th message, and $N$, the sum of every $n_i$. So, from the perspective of the A.N., $C_1$ is charged per message, $C_2$ is charged per message size, and $C_3$ works as a base fee.

$$ C_{block}(m) = \sum_{i=1}^m C_{tx}(n_i,m) = m \cdot C_1 + N \cdot C_2 + C_3 $$

To extend the formula for cases where no message has arrived ($m=0$) before CTI, let $C_{block}(0)=C_3$. Also, $C_3$ should be large enough so that the A.N. is still incentivized to publish an empty block to the blockchain.

**Reply by augustoteixeira** [reply]
I like the formula, because it closely reflects the actual cost of submitting the txs. The only problem is that the cost will show as very high to the first people in the block, because they will be paying the flat ~40k gas to initiate the tx on their own. Even though in the end they will pay less, because more people got in later on.

Here is an idea to mitigate this problem. We could finish an aggregated block without including all the txs that have been submitted. For example, if the aggregator has 100 txs, they could submit a block with 60. Now whoever arrives and submits an extra tx will already see a reduced cost (because they will go together with the 40 that were left. This may require triple time travel though. @felipeargento?

**Reply by felipeargento** [reply]
This is an interesting discussion. I see two approaches that we could consider:

1) Fee auction - similar to Ethereum. There is a crucial distinction here, because on Ethereum the validators are paid regardless of the existence of transaction, with ether emission. In our case, the validators would risk running at a loss - unless we find a way for them to be rewarded independently of the number of transaction in a block, using the DApps (or the macro economy). But perhaps the PoS should be the one covering for this blockchain interaction fee. This might only makes sense if we're considering that the aggregators can't include "infinite" inputs in the same aggregated block, otherwise the market fee wouldn't work properly.

2) The approach you guys are debating, where aggregators would never run at a loss. In the beginning though, this might make it more expensive to use the aggregator than to send direct inputs for a while, before demand picks up. 

I like the formula @guidanoli proposed and I think here we could do something like a nth price auction with the fee. The first input to arrive pays a `max fee / byte` - if that is above breakeven for the next block, the aggregator tells you that your transaction will surely be included. If not, the aggregator will tell you how many transactions they need to receive in order for the block to be profitable (and then you can choose to bump that fee). In the end, all transactions pay the smallest proportional fee that made it into that block. In this scenario users that are not willing to pay for the inclusion cost by themselves will understand that their input might not be included in the next block.

@augustoteixeira This is an interesting idea...but I am afraid that we end up hindering predictability a bit too much for the direct input. The aggregator is still earning 100 payments for two interactions, regardless of how they divide them (99 + 1 our 60 + 40). I agree that the cost would be more well divided between the users, but it still sounds a bit odd. 
I wonder if it would be better to get rid of this requirement of aggregators being profitable from the fees alone for every single block. They still have the PoS rewards and the AEV to attenuate the unprofitable periods. What do you guys think?

**Reply by augustoteixeira** [reply]
I find it risky to trust the PoS reward to pay for large blocks whose fees have been miscalculated.

**Comment by guidanoli** [comment]
> Is the transaction’s arriving time the only parameter to order transactions (i.e., FIFO)?

If we allow users to specify how much they are willing to pay for their tx to be included in an aggregated tx, then could this be another parameter to order transactions?

**Comment by augustoteixeira** [comment]
> How to collect and calculate fees? Do all transactions pay the same fee?

I am starting to think that aggregation should be rolled out together with the universal wallet. This would in particular make the payment for aggregation a lot simpler.

> Is the transaction’s arriving time the only parameter to order transactions (i.e., FIFO)?

I would say so. FIFO.

> If an aggregator is not a sequencer, could instant finality be guaranteed? I guess not.

I didn't understand the question. You are asking if the aggregator is also the one ordering the transactions? I would say so. If we pair this with double time travel we get instant feedback (which is not the same as instant finality).

> Should we consider a minimum number of transactions to be aggregated?

I would say no. We post the aggregated blob periodically and if there is a single tx, we post it with time-travel guarantees.

> Is it possible for the sender to cancel a transaction during the collection period?

I would say, no. Because people may have already received feedback including the processing of that tx.

> Is there any restriction for a dApp/user to send transactions? Do we need an authentication process?

I would say no. The Dapp is responsible to validate tx inside the linux machine. The data availability fees should be enough to prevent a DoS attack where the transactions are immediately dropped.

> What happens if an aggregator goes offline?

While this is a centralized process: They get punished financially, the transactions are not submitted, feedback may have been wrong.

> What happens if the aggregator’s output is rejected?

I think that in aggregation mode, inputs should not be reverted (unless perhaps the signature does not match).
